### PR TITLE
CODEOWNERS protocol buffer

### PR DESCRIPTION
**Pull request by cbart**
This change:

- Adds a codeowners.proto file which structurally represents a CODEOWNERS file in a way that should work with GitHub, GitLab and Bitbucket Cloud.
- `go generate buf` will now generate golang bindings from every directory with `buf.gen.yaml` present.
- Some aspects of the configuration, like optional sections are omitted, as they are not relevant to discerning code ownership.

## Test plan

Manually used code-gen to generate Go proto bindings.

For the changes to the `sg generate` command:

```
➜ cd dev/sg
➜  sg git:(cbart/codeowners-proto) ✗ PGUSER=sourcegraph go run . -config ../../sg.config.yaml  generate
👉 Running target "buf" (Re-generate protcol buffer bindings using buf)
ℹ️ goimports -w
✅ Target "buf" done (4s)
👉 Running target "go" (Run go generate [packages...] on the codebase)
✅ 1/1 packages generated 
ℹ️ go generate ./... (excluding doc/cli/references)
ℹ️ goimports -w
ℹ️ go mod tidy
✅ Target "go" done (4s)
```

### CODE REVIEW

File: internal/own/codeowners/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Codeowners is represented in an abstract manner using a Protocol Buffer structure.
+
+If making changes to the proto file, please refresh the generated Go code by running:
+
+```
+./internal/own/codeowners/proto-generate.sh
```
Comment (eseliger) [inline internal/own/codeowners/README.md:6]: should we add this to `sg generate`? Then regressions will be caught in CI

File: internal/own/codeowners/proto/codeowners.proto [hunk 2]
```diff
@@ -0,0 +1,82 @@
+// Representation of code ownership for a repository as described in a CODEOWNERS file.
+// As various orgainzations have slightly different syntax for CODEOWNERS files,
```
Comment (eseliger) [inline internal/own/codeowners/proto/codeowners.proto:2]:
```suggestion
// As various implementations have slightly different syntax for CODEOWNERS files,
```

File: internal/own/codeowners/proto/codeowners.proto [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    //   rooted at the directory root, for `/src/dir/*`:
+    //   * `/src/dir/file` matches.
+    //   * `/main/src/dir/file` does not match, as `src` is not top-level.
+    //   * `/src/dir/another/file` does not match as `*` matches
+    //     only files directly contained in specified directory.
+    // * In the above patterns `/**/` can be used to match any sub-path
+    //   between two parts of a pattern. For example: `/docs/**/internal/`
+    //   will match `/docs/foo/bar/internal/file`.
+    // * The file part of the pattern can use a `*` wildcard like so:
+    //   `docs/*.md` will match `/src/docs/index.md` but not `/src/docs/index.js`.
```
Comment (eseliger) [inline internal/own/codeowners/proto/codeowners.proto:49]: should we add a section about negation rules as bitbucket uses them?

**Review by eseliger — APPROVED** [review]

File: internal/own/codeowners/proto/codeowners.proto [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    // when evaluating owners, the result also contains a separate
+    // owners for the PM section.
+    string section_name = 3;
+}
+
+// Owner is denoted by either a handle or an email.
+// We expect exactly one of the fields to be present.
+message Owner {
+    // Handle can refer to a user or a group defined externally.
+    // In the text config, a handle always starts with `@`.
```
Comment (eseliger) [inline internal/own/codeowners/proto/codeowners.proto:75]:
```suggestion
    // Handle can refer to a user or a team defined externally.
    // In the text config, a handle always starts with `@`.
```

File: internal/own/codeowners/proto/codeowners.proto (hunk 3 above)
Comment (cbart) [reply internal/own/codeowners/proto/codeowners.proto:49]: 👍

File: internal/own/codeowners/README.md (hunk 1 above)
Comment (cbart) [reply internal/own/codeowners/README.md]: 👍

**Comment by cbart** [comment]
Adding @efritz who, I believe worked on `sg` quite a lot. Eric, could you advice if the changes to `sg generate` I made here to generate golang bindings for protobuf make sense? 🙃

**Review by mrnugget — APPROVED** [review]

File: dev/sg/generates.go [hunk 5]
```diff
@@ -5,10 +5,26 @@ import (
...
 	"github.com/sourcegraph/sourcegraph/dev/sg/internal/generate"
 	"github.com/sourcegraph/sourcegraph/dev/sg/internal/generate/golang"
+	"github.com/sourcegraph/sourcegraph/dev/sg/internal/generate/proto"
 	"github.com/sourcegraph/sourcegraph/dev/sg/root"
 )
 
 var allGenerateTargets = generateTargets{
+	{
+		// Protocol Buffer generation runs before go, as otherwise
+		// go mod tidy errors on generated protobuf go code direstories.
```
Comment (jhchabran) [inline dev/sg/generates.go:15]:
```suggestion
		// go mod tidy errors on generated protobuf go code directories.
```

File: dev/sg/internal/generate/proto/proto.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	"github.com/sourcegraph/sourcegraph/dev/sg/internal/std"
+	"github.com/sourcegraph/sourcegraph/dev/sg/root"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+var dependencies = []dependency{
+	"github.com/bufbuild/buf/cmd/buf",
+	"github.com/pseudomuto/protoc-gen-doc/cmd/protoc-gen-doc",
+	"golang.org/x/tools/cmd/goimports",
+	"google.golang.org/protobuf/cmd/protoc-gen-go",
```
Comment (jhchabran) [inline dev/sg/internal/generate/proto/proto.go:25]: Don't we want to pin a version for those? Without we could break the CI on an update of one of the binaries, if they were to include breaking changes (with the exception of `goimports`).

File: dev/sg/internal/generate/proto/proto.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+	for _, p := range bufGenFilePaths {
+		dir := filepath.Dir(p)
+		os.Chdir(dir)
+		if err := runBufGenerate(ctx, gobin, &sb); err != nil {
+			return &generate.Report{Err: err}
+		}
+	}
+
+	// Run goimports -w
+	if err := golang.RunGoImports(ctx, golang.NormalOutput, reportOut, &sb); err != nil {
```
Comment (jhchabran) [inline dev/sg/internal/generate/proto/proto.go:77]: Just wondering, why do we need to run this? I checkout the branch and it runs fine without it + `sg generate go` doesn't complain either.

File: internal/own/codeowners/proto/codeowners.pb.go [hunk 8]
```diff
@@ -0,0 +1,367 @@
+// Representation of code ownership for a repository as described in a CODEOWNERS file.
+// As various orgainzations have slightly different syntax for CODEOWNERS files,
```
Comment (jhchabran) [inline internal/own/codeowners/proto/codeowners.pb.go:2]: @cbart This has not been regenerated, this will fail the CI somewhere down the line 😊

**Review by jhchabran — CHANGES_REQUESTED** [review]
> Adding @efritz who, I believe worked on sg quite a lot. Eric, could you advice if the changes to sg generate I made here to generate golang bindings for protobuf make sense? 🙃

Took the liberty of doing it as I've played with this a lot, brought forward a few small things :)

File: dev/sg/internal/generate/proto/proto.go (hunk 6 above)
Comment (cbart) [reply dev/sg/internal/generate/proto/proto.go]: Good spot!

File: dev/sg/internal/generate/proto/proto.go (hunk 7 above)
Comment (cbart) [reply dev/sg/internal/generate/proto/proto.go]: 🤷 good call, I was wondering too if it's necessary, but saw it in so many places using `buf` online, that I did it. Let's drop it from here (makes the diff smaller), and add it later if there is actually a reason to do so.

File: internal/own/codeowners/proto/codeowners.pb.go (hunk 8 above)
Comment (cbart) [reply internal/own/codeowners/proto/codeowners.pb.go]: That's right! Fixed now. Thanks!

**Review by cbart — COMMENTED** [review]
Thanks!

**Review by jhchabran — APPROVED** [review]

**Pull request merged**